raft.go

```diff
@@ -1684,6 +1703,24 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 }
 
 func (r *raft) handleHeartbeat(m pb.Message) {
+	if r.allowInvariantViolations {
+		// If the following commit panics, it means the commit id in the message is rejected by the
+		// log since it does not have those entries. Hitting this panic usually means that a follower
+		// has not upheld its durability guarantees. If allowInvariantViolations is enabled, we take
+		// a naive approach to recover from the panic: send a pb.MsgHeartbeatResp with a reject hint,
+		// forcing the leader to retransmit the entries.
+		defer func() {
+			if v := recover(); v != nil {
```

Comment:
imo adding `recover` adds more complexity. https://github.com/etcd-io/raft/pull/25#issuecomment-1449055381 suggested different approach: `default logger would panic but users could define a logger that would just log the event and keep going`